serialization/class_definition.go

```diff
@@ -0,0 +1,177 @@
+}
+
+func (cd *ClassDefinition) AddInt64Field(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeInt64)
+}
+
+func (cd *ClassDefinition) AddFloat32Field(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeFloat32)
+}
+
+func (cd *ClassDefinition) AddFloat64Field(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeFloat64)
+}
+
+func (cd *ClassDefinition) AddStringField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeString)
+}
+
+func (cd *ClassDefinition) AddPortableField(fieldName string, def *ClassDefinition) error {
+	if def.ClassID == 0 {
+		return hzerrors.NewHazelcastIllegalArgumentError("Portable class ID cannot be zero", nil)
+	}
+	cd.AddField(NewFieldDefinition(int32(len(cd.Fields)), fieldName, TypePortable, def.FactoryID, def.ClassID, cd.Version))
+	return nil
+}
+
+func (cd *ClassDefinition) addNewFieldDefinition(fieldName string, fieldType int32) {
+	cd.AddField(NewFieldDefinition(int32(len(cd.Fields)), fieldName, fieldType, 0, 0, cd.Version))
+}
+
+func (cd *ClassDefinition) AddByteArrayField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeByteArray)
+}
+
+func (cd *ClassDefinition) AddBoolArrayField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeBoolArray)
+}
+
+func (cd *ClassDefinition) AddInt16ArrayField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeInt16Array)
+}
+
+func (cd *ClassDefinition) AddUInt16ArrayField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeUInt16Array)
+}
+
+func (cd *ClassDefinition) AddInt32ArrayField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeInt32Array)
+}
+
+func (cd *ClassDefinition) AddInt64ArrayField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeInt64Array)
+}
+
+func (cd *ClassDefinition) AddFloat32ArrayField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeFloat32Array)
+}
+
+func (cd *ClassDefinition) AddFloat64ArrayField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeFloat64Array)
+}
+
+func (cd *ClassDefinition) AddPortableArrayField(fieldName string, def *ClassDefinition) error {
+	if def.ClassID == 0 {
+		return hzerrors.NewHazelcastIllegalArgumentError("Portable class ID cannot be zero", nil)
+	}
+	cd.AddField(NewFieldDefinition(int32(len(cd.Fields)), fieldName, TypePortableArray, def.FactoryID, def.ClassID, cd.Version))
+	return nil
+}
+
+func (cd *ClassDefinition) AddStringArrayField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeStringArray)
+}
+
+const (
+	TypePortable int32 = iota
+	TypeByte
+	TypeBool
+	TypeUint16
+	TypeInt16
+	TypeInt32
+	TypeInt64
+	TypeFloat32
+	TypeFloat64
+	TypeString
+	TypePortableArray
+	TypeByteArray
+	TypeBoolArray
+	TypeUInt16Array
+	TypeInt16Array
+	TypeInt32Array
+	TypeInt64Array
+	TypeFloat32Array
+	TypeFloat64Array
+	TypeStringArray
+)
+
+// FieldDefinition defines name, type, index of a field.
+type FieldDefinition struct {
+	Name      string
```

Comment:
is there a way to make the struct fields immutable, without sacrificing their simplicity to access in Go?

Reply:
There's nothing like read-only fields. But there's no need to make `FieldDefinition` immutable / limit access, since it gets copied before use. User modifications after passing the configuration aren't visible to the internals.